src/component_selector.jl

```diff
@@ -0,0 +1,408 @@
+One MAY subtype this and have a `groupby::Union{Symbol, Function}` field to get an automatic
+implementation of `get_groups`.
+=#
+"`PluralComponentSelector`s whose grouping is determined by a `groupby` field."
+abstract type DynamicallyGroupedComponentSelector <: PluralComponentSelector end
+
+# COMMON COMPONENTSELECTOR INFRASTRUCTURE
+"Canonical way to turn an `InfrastructureSystemsComponent` subtype into a unique string."
+subtype_to_string(subtype::Type{<:InfrastructureSystemsComponent}) =
+    strip_module_name(subtype)
+
+"""
+Canonical way to turn an `InfrastructureSystemsComponent` specification/instance into a
+unique-per-system string.
+"""
+component_to_qualified_string(
+    component_type::Type{<:InfrastructureSystemsComponent},
+    component_name::AbstractString,
+) = subtype_to_string(component_type) * COMPONENT_NAME_DELIMITER * component_name
+component_to_qualified_string(component::InfrastructureSystemsComponent) =
+    component_to_qualified_string(typeof(component), get_name(component))
+
+const VALID_GROUPBY_KEYWORDS = [:all, :each]
+
+# Ideally we could leave system-like arguments untyped for maximum extensibility, but
+# because only PSY.get_components, not IS.get_components, is defined for PSY.System, we need
+# to be able to easily override these methods. See
+# https://github.com/NREL-Sienna/InfrastructureSystems.jl/issues/388
+const SystemLike = Union{Components, SystemData}
+
+"""
+Helper function to check that the `groupby` argument is valid. Passes it through if so,
+errors if not.
+"""
+validate_groupby(groupby::Symbol) =
+    if (groupby in VALID_GROUPBY_KEYWORDS)
+        groupby
+    else
+        throw(ArgumentError("groupby must be one of $VALID_GROUPBY_KEYWORDS or a function"))
+    end
+validate_groupby(groupby::Function) = groupby  # Don't try to validate functions for now
+
+# Generic implementations/generic docstrings for simple functions with many methods
+"""
+Factory function to create the appropriate subtype of `ComponentSelector` given the
+arguments. Users should call this rather than manually constructing `ComponentSelector`
+subtypes.
+"""
+function make_selector end
+
+# Override this if you define a ComponentSelector subtype with no name field
+"""
+Get the name of the `ComponentSelector`. This is either the default name or a custom name
+passed in at creation time.
+"""
+get_name(selector::ComponentSelector) = selector.name
+
+"""
+    get_components(selector, sys; scope_limiter = nothing)
+Get the components of the collection that make up the `ComponentSelector`.
+ - `scope_limiter`: optional filter function to limit the scope of components under consideration
+"""
+function get_components end
+
+"""
+    get_components(scope_limiter, selector, sys)
+Get the components of the collection that make up the `ComponentSelector`.
+ - `scope_limiter`: optional filter function to limit the scope of components under consideration
+"""
+get_components(
+    scope_limiter::Union{Nothing, Function},
```

Comment:
It is mostly a stylistic thing, I could concoct a scenario where the user is broadcasting/looping and splatting where it would make sense (when you get to the PowerAnalytics PR you'll see some cases where this does actually happen) but that does seem unlikely in this case

Reply:
Resolving as obsolete: `scope_limiter` is no longer part of the user interface.